### DISCUSSION TITLE
How to update JSON field?

### DISCUSSION THREAD

**Discussion by mationai** [supabase/supabase#447] [Questions] [answered]
I have a 'group' table with 'name' text field and 'lists' jSON. I have a row with name: 'groupname' and lists: ['alist'] (shows as `{"alist"}` in table editor.

I'm getting:
```
{"details":"unexpected \"[\" expecting field name (* or [a..z0..9_])","message":"\"failed to parser tree path ([object Object])\" (line 1, column 1)"}
```
error when trying to do:
```javascript
    await supa.from('group').update({ lists: ['a']}).eq({ name: 'groupname' })
```
Changing ['a'] to JSON.stringify(['a']) didn't work either.

**Comment by soedirgo** [comment]
Hey @fuzzthink!

The syntax for `eq` is like so:

```js
await supa.from('group').update({ lists: JSON.stringify(['a'])}).eq('name', 'groupname') // Yep, you do need to stringify it
```

Alternatively, you can use `match` if you want to `eq` on multiple columns (see [here](https://supabase.io/docs/client/match#with-update)).

Let me know if this helps :).

**Reply by soedirgo** [reply]
Btw, you might get this error when modifying tables with no primary key: `cannot update table \"group\" because it does not have a replica identity and publishes updates`. In this case just run this in the SQL editor:

```sql
ALTER TABLE "group" REPLICA IDENTITY FULL;
```

We're working on making it so that you don't have to do this manually.

**Comment by mationai** [comment, marked as answer]
@soedirgo 
Thank you for pointing out the correct syntax for `.eq` -- which btw, is `.eq('colname', 'val')` (',' instead of  ':').

Trying to update with `JSON.stringify(['new', 'hot'])`, I'm getting:

`{"hint":null,"details":"\"[\" must introduce explicitly-specified array dimensions.","code":"22P02","message":"malformed array literal: \"[\"new\",\"hot\"]\""}`

**Reply by mationai** [reply]
@soedirgo Any thoughts on how to fix the above error? Thanks.

**Reply by soedirgo** [reply]
I wasn't able to reproduce this. Can you show me the code? Ftr this is what I did:
```js
const { data } = await supabase
    .from("group")
    .update({ lists: JSON.stringify(["new", "hot"]) })
    .eq("name", "groupname");
```

**Reply by mationai** [reply]
It appears that JSON fields should be passed an actual object, not a string, as the error no longer appears and writes correctly after removing `JSON.stringify()`.

For reference, the solution should be:

```javascript
await supabase
    .from("tablename")
    .update({ jsonColname: ["a", "b"] })
    .eq("colnameToMatch", 'val') // or if match multiple cols:
//  .match({"colnameToMatch": 'val', "col2ToMatch": 'val'})
```

**Reply by soedirgo** [reply]
Ah, got it - actually it works for me either with or without `JSON.stringify`. Glad this works for you!

**Comment by jarretmoses** [comment]


**Reply by jarretmoses** [reply]
seemingly it is a policy issue but I created a policy for update

**Comment by mousindev** [comment]
Hi. I am facing a different situation and can't find a way in the documentation to achieve it. In my case, the json fields are objects, not an array, where I store key values for language code and the corresponding translation. So, if the column is `translations`, the object would be like: `{ "en": "Default Text", "es": "Texto por defecto"}`. The thing is that the user can decide what languages will be part of the translations column (even though I don't know it upfront I can `[]` the language code to get it as a key inside the json object. How can I only edit one of the languages / keys? So far, if I try with the `update()` method as follows the full column value gets overwritten: 
```js
    ...
    .update(
        translations: {
            [language_code]: translation_value
        }
    );
```
so, if the code above would have "es" as language code and "Nuevo Texto por Defecto " as value, the english text would be lost.

Thanks

**Reply by 3den** [reply]
I have exactly the same problem, the way I solving it is hacky:

```js
    ...
    .update({
        translations: {
            ...item.translations, // im getting current item first to merge the existing jsonb with the new one
            [language_code]: translation_value
        }
    });
```

I would love to know if there is a better way to do that.

**Reply by squidgemonster** [reply]
I'd be very curious about this too. The Supabase docs do say 'Currently, it is only possible to update the entire JSON document.' (see Notes on https://supabase.com/docs/reference/javascript/update?example=updating-json-data) - but it's a bit of a pain having to grab the full JSON to just update one value...
